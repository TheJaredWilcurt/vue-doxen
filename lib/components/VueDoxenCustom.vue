<template>
  <div v-bind="applyStyleTokens({ vueDoxen: true })">
    <ComponentDemo
      v-if="processedDemos[selectedDemo]"
      :demo="processedDemos[selectedDemo]"
      :options="validatedOptions"
      :styleTokens="styleTokens"
      :key="selectedDemo"
    />
  </div>
</template>

<script>
/**
 * @file This component acts as a wrapper and public interface for ComponentDemo.
 * It handles the user inputs, defaulting values, and passing down just the selected demo.
 */
import {
  createImportStatement,
  processDemos
} from '@/helpers/componentHelpers.js';
import {
  createModelValueProp,
  createVueDoxenOptions,
  demos,
  styleTokens
} from '@/helpers/props.js';
import { validateOptions } from '@/helpers/validateOptions.js';

import applyStyleTokens from '@/mixins/applyStyleTokensMixin.js';

import ComponentDemo from '@/components/ComponentDemo.vue';
import VueDoxenCustomDescription from '@/components/VueDoxenCustomDescription.vue';

const COMPONENT_NAME = 'VueDoxenCustom';
const modelValue = createModelValueProp(String);
const options = createVueDoxenOptions(false);

export default {
  ...createImportStatement(COMPONENT_NAME),
  name: COMPONENT_NAME,
  description: {
    component: VueDoxenCustomDescription
  },
  components: {
    ComponentDemo
  },
  mixins: [
    applyStyleTokens
  ],
  emits: ['update:model-value'],
  props: {
    demos,
    modelValue,
    options,
    styleTokens
  },
  methods: {
    defaultModelValue: function () {
      if (
        this.processedDemos &&
        Object.keys(this.processedDemos).length &&
        !Object.keys(this.processedDemos).includes(this.modelValue)
      ) {
        const firstDemo = Object.keys(this.processedDemos)[0];
        console.info(
          'Vue-Doxen received a v-model value of "' + this.modelValue +
          '", however that value could not be found in the list of ' +
          'provided demos:', Object.keys(this.processedDemos)
        );
        console.info(
          'This is likely from a bug in your code, that you should ' +
          'fix. For convenience we will display the first provided ' +
          'demo, "' + firstDemo + '".'
        );
        this.updateValue(firstDemo);
      }
    },
    updateValue: function (key) {
      this.$emit('update:model-value', key);
    }
  },
  computed: {
    selectedDemo: function () {
      if (
        this.processedDemos &&
        Object.keys(this.processedDemos).length
      ) {
        if (Object.keys(this.processedDemos).includes(this.modelValue)) {
          return this.modelValue;
        }
        return Object.keys(this.processedDemos)[0];
      }
      return undefined;
    },
    processedDemos: function () {
      return processDemos(this.demos);
    },
    validatedOptions: function () {
      return validateOptions(this.options);
    }
  },
  watch: {
    modelValue: function () {
      this.defaultModelValue();
    }
  },
  created: function () {
    this.defaultModelValue();
  }
};
</script>
