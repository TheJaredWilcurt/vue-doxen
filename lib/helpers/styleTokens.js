/**
 * @file Style tokens allow a lot of flexibility in how the Vue-Doxen components can be styled.
 *
 * Style tokens solve the following scenarios:
 *
 * Provide optional basic built in CSS
 * Allow overriding basic built-in classes
 * The CSS is externalized and can be skipped if not needed
 * You can remove all classes from all elements if not needed
 * Allows customizing the class names to whatever you want
 * Allows having many classes applied on a single element (Tailwind)
 * Can use existing CSS class names (Bootstrap)
 * Can use existing CSS component classes (Vuetify)
 */

/**
 * A predictable map of tokens to class names. These classes are not used anywhere, they exist to make styling easier for those writing their own CSs.
 *
 * @type {Object}
 */
export const styleTokensBuiltIn = {
  button: 'doxen-button',
  buttonNotSelected: 'doxen-button-not-selected',
  buttonSelected: 'doxen-button-selected',
  codeBox: 'doxen-code-box',
  codeSwapper: 'doxen-code-swapper',
  codeSwapperFileName: 'doxen-code-swapper-file-name',
  componentDemo: 'doxen-component-demo',
  componentDemoContainer: 'doxen-component-demo-container',
  componentDemoH3: 'doxen-component-demo-h3',
  componentDemoHr: 'doxen-component-demo-hr',
  demoDescription: 'doxen-demo-description',
  demoHeader: 'doxen-demo-header',
  demoHeaderTitle: 'doxen-demo-header-title',
  emitsDocumentationCode: 'doxen-emits-documentation-code',
  emitsDocumentationLi: 'doxen-emits-documentation-li',
  emitsDocumentationSpan: 'doxen-emits-documentation-span',
  emitsDocumentationStrong: 'doxen-emits-documentation-strong',
  emitsDocumentationUl: 'doxen-emits-documentation-ul',
  emitLogClearButton: 'doxen-emit-log-clear-button',
  emitLogHeader: 'doxen-emit-log-header',
  emitLogTable: 'doxen-emit-log-table',
  emitLogTableContainer: 'doxen-emit-log-table-container',
  emitLogTbody: 'doxen-emit-log-tbody',
  emitLogTd: 'doxen-emit-log-td',
  emitLogTh: 'doxen-emit-log-th',
  emitLogThead: 'doxen-emit-log-thead',
  emitLogTr: 'doxen-emit-log-tr',
  formFieldCheckbox: 'doxen-form-field-checkbox',
  formFieldCheckboxContainer: 'doxen-form-field-checkbox-container',
  formFieldCheckboxError: 'doxen-form-field-checkbox-error',
  formFieldCheckboxNameDisabled: 'doxen-form-field-checkbox-name-disabled',
  formFieldCheckboxNameError: 'doxen-form-field-checkbox-name-error',
  formFieldCheckboxNameLabel: 'doxen-form-field-checkbox-name-label',
  formFieldDropdown: 'doxen-form-field-dropdown',
  formFieldDropdownContainer: 'doxen-form-field-dropdown-container',
  formFieldDropdownError: 'doxen-form-field-dropdown-error',
  formFieldDropdownOption: 'doxen-form-field-dropdown-option',
  formFieldError: 'doxen-form-field-error',
  formFieldFieldset: 'doxen-form-field-fieldset',
  formFieldLabel: 'doxen-form-field-label',
  formFieldLegend: 'doxen-form-field-legend',
  formFieldLegendDisabled: 'doxen-form-field-legend-disabled',
  formFieldLegendError: 'doxen-form-field-legend-error',
  formFieldMessage: 'doxen-form-field-message',
  formFieldRadioDial: 'doxen-form-field-radio-dial',
  formFieldRadioDialContainer: 'doxen-form-field-radio-dial-container',
  formFieldRadioDialNameDisabled: 'doxen-form-field-radio-dial-name-disabled',
  formFieldRadioDialNameError: 'doxen-form-field-radio-dial-name-error',
  formFieldRadioDialNameLabel: 'doxen-form-field-radio-dial-name-label',
  formFieldRadioDialsContainer: 'doxen-form-field-radio-dials-container',
  formFieldRequired: 'doxen-form-field-required',
  formFieldTextarea: 'doxen-form-field-textarea',
  formFieldTextareaCode: 'doxen-form-field-textarea-code',
  formFieldTextareaContainer: 'doxen-form-field-textarea-container',
  formFieldTextareaError: 'doxen-form-field-textarea-error',
  formFieldTextField: 'doxen-form-field-text-field',
  formFieldTextFieldContainer: 'doxen-form-field-text-field-container',
  formFieldTextFieldError: 'doxen-form-field-text-field-error',
  propsDocumentationCode: 'doxen-props-documentation-code',
  propsDocumentationEm: 'doxen-props-documentation-em',
  propsDocumentationLi: 'doxen-props-documentation-li',
  propsDocumentationSpan: 'doxen-props-documentation-span',
  propsDocumentationStrong: 'doxen-props-documentation-strong',
  propsDocumentationUl: 'doxen-props-documentation-ul',
  propsPlaygroundForm: 'doxen-props-playground-form',
  sidebarButton: 'doxen-sidebar-button',
  sidebarButtonNotSelected: 'doxen-sidebar-button-not-selected',
  sidebarButtonSelected: 'doxen-sidebar-button-selected',
  sidebarContainer: 'doxen-sidebar-container',
  tabButton: 'doxen-tab-button',
  tabButtonNotSelected: 'doxen-tab-button-not-selected',
  tabButtonSelected: 'doxen-tab-button-selected',
  tabsButtonContainer: 'doxen-tabs-button-container',
  tabsHeader: 'doxen-tabs-header',
  vueDoxen: 'doxen-vue-doxen'
};

/**
 * All the style token keys without any class names applied to them.
 *
 * @type {Object}
 */
export const styleTokensEmpty = {};
for (const key in styleTokensBuiltIn) {
  styleTokensEmpty[key] = '';
}

/**
 * Style tokens using the class names provided by Bootstrap 5.
 *
 * @type {Object}
 */
export const styleTokensBootstrap5 = {
  // Container
  formFieldFieldset: 'px-2 pb-2 m-1 flex-grow-1',
  // Errors
  formFieldCheckboxError: 'text-danger',
  formFieldCheckboxNameError: 'text-danger',
  formFieldDropdownError: 'text-danger',
  formFieldError: 'form-text text-danger',
  formFieldLegendError: 'text-danger',
  formFieldRadioDialNameError: 'text-danger',
  formFieldTextareaError: 'text-danger',
  formFieldTextFieldError: 'text-danger',
  // Labels
  formFieldLegend: 'form-label form-control-sm mb-0 px-0',
  // Inputs
  formFieldTextarea: 'form-control',
  formFieldTextField: 'form-control',
  // Checkbox
  formFieldCheckbox: 'form-check-input',
  formFieldCheckboxContainer: 'form-check',
  formFieldCheckboxNameDisabled: '',
  formFieldCheckboxNameLabel: 'form-check-label',
  // Radio Dials
  formFieldRadioDial: 'form-check-input',
  formFieldRadioDialNameLabel: 'form-check-label form-label form-control-sm',
  formFieldRadioDialContainer: 'form-check',
  // Select
  formFieldDropdown: 'form-select form-select-sm',
  // Message
  formFieldMessage: 'form-text',
  // Buttons
  button: 'btn',
  buttonSelected: 'btn-primary',
  buttonNotSelected: 'btn-outline-primary',
  // Code
  codeSwapperFileName: 'float-end',
  formFieldTextareaCode: 'font-monospace',
  tabsButtonContainer: 'nav nav-tabs nav-item mt-3 mb-0',
  tabButton: 'nav-link me-1',
  tabButtonSelected: 'active border-bottom-0 mb-0',
  // Demo
  propsPlaygroundForm: 'd-flex flex-wrap',
  // Emit log
  emitLogHeader: 'd-flex',
  emitLogClearButton: 'btn btn-primary',
  emitLogTable: 'table table-striped',
  // Sidebar
  sidebarContainer: 'btn-group-vertical mx-3 justify-content-start',
  sidebarButton: 'btn flex-grow-0',
  sidebarButtonNotSelected: 'btn-outline-primary',
  sidebarButtonSelected: 'btn-primary flex-shrink-1'
};
