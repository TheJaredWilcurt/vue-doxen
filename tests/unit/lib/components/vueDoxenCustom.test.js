import { markRaw } from 'vue';

import { styleTokensBuiltIn } from '@/helpers/styleTokens.js';

import VueDoxenCustom from '@/components/VueDoxenCustom.vue';

import testHelpers from '@@/unit/testHelpers.js';

describe('VueDoxenCustom.vue', () => {
  const demos = {
    DummyButton: markRaw({
      name: 'DummyButton',
      description: 'Dummy button description',
      template: '<button>DummyButton</button>'
    })
  };
  const modelValue = 'DummyButton';
  function makeDummyComponent (name) {
    return markRaw({
      name: 'Dummy' + name,
      template: '<div>Dummy' + name + '</div>'
    });
  }
  const options = {
    components: {
      checkbox: makeDummyComponent('Checkbox'),
      dropdown: makeDummyComponent('Dropdown'),
      emitLog: makeDummyComponent('EmitLog'),
      emitsDocumentation: makeDummyComponent('EmitsDocumentation'),
      header: makeDummyComponent('Header'),
      jsonTextarea: makeDummyComponent('JsonTextarea'),
      numberField: makeDummyComponent('NumberField'),
      plainText: makeDummyComponent('PlainText'),
      propsDocumentation: makeDummyComponent('PropsDocumentation'),
      radioDials: makeDummyComponent('RadioDials'),
      rangeSlider: makeDummyComponent('RangeSlider'),
      textField: makeDummyComponent('TextField'),
      textarea: makeDummyComponent('Textarea')
    }
  };
  const styleTokens = {
    ...styleTokensBuiltIn,
    vueDoxen: {
      class: 'doxen-vue-doxen-test',
      'data-example': 'test'
    }
  };
  const requiredProps = {
    options
  };
  let consoleInfo;

  beforeEach(() => {
    consoleInfo = console.info;
    console.info = vi.fn();
  });

  afterEach(() => {
    console.info = consoleInfo;
  });

  const setupWrapper = async (props) => {
    const componentOptions = {
      props: {
        ...requiredProps,
        ...props
      }
    };
    const wrapper = await testHelpers.mount(VueDoxenCustom, componentOptions);
    return wrapper;
  };

  test('Renders correctly with required props', async () => {
    const wrapper = await setupWrapper();

    expect(console.info)
      .not.toHaveBeenCalled();

    expect(wrapper)
      .toMatchSnapshot();
  });

  test('Renders correctly with all props', async () => {
    const props = {
      demos,
      modelValue,
      styleTokens
    };
    const wrapper = await setupWrapper(props);

    expect(console.info)
      .not.toHaveBeenCalled();

    global.vueSnapshots.stringifyAttributes = true;

    expect(wrapper)
      .toMatchSnapshot();
  });

  test('Defaults to first demo if v-model is a mismatch', async () => {
    const props = {
      demos,
      modelValue: 'MISMATCH',
      styleTokens
    };
    const wrapper = await setupWrapper(props);

    expect(console.info)
      .toHaveBeenCalledWith(
        'Vue-Doxen received a v-model value of "MISMATCH", ' +
        'however that value could not be found in the list of provided demos:',
        ['DummyButton']
      );

    expect(console.info)
      .toHaveBeenCalledWith(
        'This is likely from a bug in your code, that you should fix. ' +
        'For convenience we will display the first provided demo, "DummyButton".'
      );

    expect(wrapper.vm.selectedDemo)
      .toEqual('DummyButton');

    expect(wrapper.emitted()['update:model-value'][0])
      .toEqual(['DummyButton']);
  });

  test('Logs warning if missing custom component and uses FallBack', async () => {
    const props = {
      demos,
      modelValue,
      options: {
        components: {
          ...options.components,
          checkbox: null
        }
      },
      styleTokens
    };
    const wrapper = await setupWrapper(props);

    expect(console.info)
      .toHaveBeenCalledWith(
        'REQUIRED: options.components.checkbox is missing a custom component. ' +
        'One MUST be provided. No fallback will be used and errors will occur.'
      );

    expect(wrapper.vm.validatedOptions.components.checkbox.name)
      .toEqual('FallBack');
  });
});
