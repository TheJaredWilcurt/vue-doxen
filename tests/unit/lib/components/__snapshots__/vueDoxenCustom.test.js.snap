// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VueDoxenCustom.vue > Renders correctly with all props 1`] = `
<div class="doxen-vue-doxen-test" data-applied-style-tokens="vueDoxen" data-example="test" data-style-tokens="vueDoxen">
  <div class="doxen-component-demo" data-applied-style-tokens="componentDemo" data-style-tokens="componentDemo">
    <!-- DoxenHeader -->
    <div styletokens="[object Object]" title="Dummy Button">DummyHeader</div>
    <div class="doxen-demo-description" data-applied-style-tokens="demoDescription" data-style-tokens="demoDescription">Dummy button description</div>
    <!--v-if-->
    <!-- Component being demo'd -->
    <div class="doxen-component-demo-container" data-applied-style-tokens="componentDemoContainer" data-style-tokens="componentDemoContainer">
      <hr class="doxen-component-demo-hr" data-applied-style-tokens="componentDemoHr" data-style-tokens="componentDemoHr">
      <button>DummyButton</button>
      <hr class="doxen-component-demo-hr" data-applied-style-tokens="componentDemoHr" data-style-tokens="componentDemoHr">
    </div>
    <h3 class="doxen-component-demo-h3" data-applied-style-tokens="componentDemoH3" data-style-tokens="componentDemoH3">Props Playground:</h3>
    <form class="doxen-props-playground-form" data-applied-style-tokens="propsPlaygroundForm" data-style-tokens="propsPlaygroundForm">
      <!-- Anything for Props -->
      <!-- Slots Playground -->
      <!-- DoxenEmitLog for emits -->
      <!--v-if-->
    </form>
    <div class="doxen-code-swapper" data-applied-style-tokens="codeSwapper" data-style-tokens="codeSwapper">
      <div class="doxen-tabs-header" data-applied-style-tokens="tabsHeader" data-style-tokens="tabsHeader">
        <div class="doxen-tabs-button-container" data-applied-style-tokens="tabsButtonContainer" data-style-tokens="tabsButtonContainer">
          <button aria-pressed="true" class="doxen-tab-button doxen-tab-button-selected" data-applied-style-tokens="tabButton tabButtonSelected" data-style-tokens="tabButton tabButtonNotSelected tabButtonSelected">Vue</button>
          <button aria-pressed="false" class="doxen-tab-button doxen-tab-button-not-selected" data-applied-style-tokens="tabButton tabButtonNotSelected" data-style-tokens="tabButton tabButtonNotSelected tabButtonSelected">JavaScript</button>
        </div>
      </div>
      <div aria-label="Code example" class="doxen-code-box-container" data-applied-style-tokens="codeBoxContainer" data-style-tokens="codeBoxContainer" role="textbox" tabindex="0" title="Click to copy"><pre class="doxen-code-box" data-applied-style-tokens="codeBox" data-style-tokens="codeBox"><code class="hljs xml" tabindex="0"><span class="hljs-tag"><<span class="hljs-name">DummyButton</span> /></span></code></pre>
        <!--v-if-->
      </div>
      <div aria-label="Code example" class="doxen-code-box-container" data-applied-style-tokens="codeBoxContainer" data-style-tokens="codeBoxContainer" role="textbox" style="display: none;" tabindex="0" title="Click to copy"><pre class="doxen-code-box" data-applied-style-tokens="codeBox" data-style-tokens="codeBox"><code class="hljs javascript" tabindex="0"><span class="hljs-keyword">const</span> dummyButtonProps = {};</code></pre>
        <!--v-if-->
      </div>
    </div><!-- DoxenPropsDocumentation -->
    <div propstodemo="[object Object]" styletokens="[object Object]">DummyPropsDocumentation</div>
    <!--v-if-->
  </div>
</div>
`;

exports[`VueDoxenCustom.vue > Renders correctly with required props 1`] = `
<div class data-applied-style-tokens="vueDoxen" data-style-tokens="vueDoxen">
  <!--v-if-->
</div>
`;
